File: apps/fsharp/src/MachineLearning/Chapter14/Main.fs
module MachineLearning.Chapter14.Main

open System.IO
open MachineLearning.Dataset
open MachineLearning.Chapter14.KMeans
open MachineLearning.Chapter14.MlNetKMeans
open MachineLearning.Chapter14.Spending
open MachineLearning.Chapter14.StatsKMeans

[<Literal>]
let Seed = 0

[<Literal>]
let NInit = 10

let ClusterCounts = [ 1..10 ]

[<Literal>]
let NClusters = 5

/// 2 つの結果が、同じクラスタ番号と（小数第 9 位まで）同じ SSE か
let private sameResult (a: KMeansResult) (b: KMeansResult) : bool =
    a.Labels = b.Labels && abs (a.Sse - b.Sse) < 1e-9

/// Wholesale.csv の支出額をクラスタリングし、SSE・ライブラリとの比較・クラスタごとの平均支出額を表示する
let run (print: string -> unit) : unit =
    let rows = loadSpending (Path.Combine(dataDir (), "Wholesale.csv"))
    let points = toStandardizedPoints SpendingColumns rows
    print $"データ件数: {rows.Length}（支出額 {SpendingColumns.Length} 列）"
    print $"クラスタ数ごとの SSE（初期中心 {NInit} 通りの最小値）:"
    print "クラスタ数\t自作\tML.NET（k-means++）"

    for n, sse in sseByClusterCount NInit Seed ClusterCounts points do
        print $"{n}\t{sse:F2}\t{mlNetBestSse n Seed NInit points:F2}"

    // 自作の試行と同じ初期中心を FSharp.Stats に渡し、結果が一致するかを数える
    let candidates =
        [
            for n in ClusterCounts do
                for i in 0 .. NInit - 1 -> chooseInitialCenters n (Seed + i) points
        ]

    let matched =
        candidates
        |> List.filter (fun centers -> sameResult (kmeans centers points) (kmeansWithFSharpStats centers points))
        |> List.length

    print $"同じ初期中心で FSharp.Stats と結果が一致した数: {matched} / {candidates.Length}"

    let result = kmeansWithRestarts NInit NClusters Seed points
    print ""
    print $"クラスタ数 {NClusters} のクラスタごとの件数と平均支出額:"
    print (String.concat "\t" ([ "クラスタ"; "件数" ] @ SpendingColumns))

    for summary in summarizeClusters SpendingColumns result.Labels rows do
        let means =
            SpendingColumns |> List.map (fun column -> $"{summary.Means[column]:F0}")

        print (String.concat "\t" ([ string summary.Cluster; string summary.Count ] @ means))

File: apps/scala/src/main/scala/machinelearning/chapter14/Main.scala
package machinelearning.chapter14

import java.nio.file.Paths
import java.util.Locale
import java.util.logging.{Level, Logger}
import machinelearning.dataset.DataDir

/** 卸売業者の顧客を支出額で K-means にかけ、エルボー法の SSE とクラスタごとの特徴を表示する。 */
object Main:
  private val Seed = 0L
  private val NInit = KMeans.DefaultNInit
  private val ClusterCounts = (1 to 10).toVector
  private val NClusters = 5

  // Tribuo が学習の経過を標準エラーに出すので、警告以上だけにする。
  // ロガーはガベージコレクションで設定ごと消えないように、フィールドで持ち続ける
  private val tribuoLogger = Logger.getLogger("org.tribuo")

  def run(print: String => Unit): Unit =
    tribuoLogger.setLevel(Level.WARNING)
    val x = Spending.load(Paths.get(DataDir.current(), "Wholesale.csv"))
    val columns = x.head.columns
    val points = Spending.standardize(x)

    print(s"データ件数: ${x.size}（支出額 ${columns.size} 列）")
    print(s"クラスタ数ごとの SSE（初期中心 $NInit 通りの最小値）:")
    print("クラスタ数\t自作\tTribuo（k-means++）")
    KMeans.sseByClusterCount(points, ClusterCounts, Seed, NInit).foreach { (n, sse) =>
      print(s"$n\t${format(sse, 2)}\t${format(TribuoKMeans.bestSse(points, n, Seed, NInit), 2)}")
    }

    val result = KMeans.fitWithRestarts(points, NClusters, Seed, NInit)
    print("")
    print(s"クラスタ数 $NClusters のクラスタごとの件数と平均支出額:")
    print((Vector("クラスタ", "件数") ++ columns).mkString("\t"))
    Spending.summarizeClusters(x, result.labels).foreach { summary =>
      val cells = Vector(summary.cluster.toString, summary.count.toString) ++
        columns.map(column => format(summary.means(column), 0))
      print(cells.mkString("\t"))
    }

  private def format(value: Double, digits: Int): String =
    String.format(Locale.ROOT, s"%.${digits}f", value)
